main.cpp

```diff
@@ -272,141 +218,162 @@ struct HeapA
 #include <cmath>
 #include <iostream>
 #include <functional>
-struct DoubleType;
-struct IntType;
+#include <memory>
+#include <type_traits>
+#include <limits>
 
 
-struct FloatType
+template<class NumType>
+struct Numeric
 {
-    FloatType(float nuFloat) : value( new float(nuFloat) ) { } 
-    ~FloatType()
-    {
-        delete value;
-        value = nullptr;
-    }
+    using Type = NumType;
 
-    operator float() const {return *value;}
+    Numeric(Type nuNum) : value( std::make_unique<Type>(nuNum) ) { } 
 
-    FloatType& operator+=( float rhs );
-    FloatType& operator-=( float rhs );
-    FloatType& operator*=( float rhs );
-    FloatType& operator/=( float rhs );
+    operator Type() const {return *value;}
 
-    FloatType& pow(float f);
-    FloatType& pow(const IntType& i);
-    FloatType& pow(const FloatType& f);
-    FloatType& pow(const DoubleType& d);
-
-    FloatType& apply( std::function<FloatType&(float&)> func )
+    Numeric& operator+=( Type rhs )
     {
-        if(func)
-            return func(*value);
+        *this->value += rhs;
         return *this;
     }
-    FloatType& apply( void(*funcPtr)(float&) )
+    Numeric& operator-=( Type rhs )
     {
-        if( funcPtr )
-            funcPtr(*value);
+        *this->value -= rhs;
+        return *this;
+    }
+    Numeric& operator*=( Type rhs )
+    {
+        *this->value *= rhs;
         return *this;
     }
 
-private: 
-    float* value = nullptr;
-    FloatType& powInternal(float arg);
-};
-
+    template<class DivType>
+    Numeric& operator/=( const DivType& rhs )
+    {
+        if constexpr ( std::is_same< Type, int >::value )
+        { 
+            if constexpr ( std::is_same< DivType, int >::value )
+            {
+                if ( rhs == 0 )
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+                    return *this;
+                }
+            }
+            else if ( rhs < std::numeric_limits<DivType>::epsilon() )
+            {
+                std::cout << "can't divide integers by zero!" << std::endl;
+                return *this;
+            }
+        }
+        else if ( rhs < std::numeric_limits<Type>::epsilon() )
+        {
+            std::cout << "warning: floating point division by zero!" << std::endl;
```

Comment:
Check if `abs(rhs)` is less than the epsilon. 
Right now any negative number will cause these conditionals to be evaluated as true
- [ ] revise